### PR TITLE
Allow nil openstax_uid; Don't send invalid openstax_uid to Accounts

### PR DESCRIPTION
**Pull request by Dantemss**

### CODE REVIEW

File: app/models/openstax/accounts/account.rb [hunk 1]
```diff
@@ -46,12 +49,36 @@ def has_authenticated?
...
+
+    def set_faculty_status
+      self.faculty_status ||= :no_faculty_info
+    end
+
+    def set_account_type(force: false)
+      return account_type if account_type.present? && !force
+
+      self.account_type = valid_openstax_uid? ? :remote : :local
+    end
```
Comment (jpslav) [inline app/models/openstax/accounts/account.rb]: Can `set_faculty_status` and `set_account_type` be `set_default_faculty_status` and `set_default_account_type` and can they be `protected`?

File: app/models/openstax/accounts/account.rb [hunk 2]
```diff
@@ -46,12 +49,36 @@ def has_authenticated?
...
+
     protected
 
-    def syncing_or_stubbing
+    def syncing_or_stubbing?
       syncing || OpenStax::Accounts.configuration.enable_stubbing?
     end
 
+    def should_send_updates_to_accounts?
+      !syncing_or_stubbing? && remote? && valid_openstax_uid?
```
Comment (jpslav) [inline app/models/openstax/accounts/account.rb]: what's a remote account with an invalid openstax uid?

File: app/models/openstax/accounts/account.rb (hunk 1 above)
Comment (jpslav) [inline app/models/openstax/accounts/account.rb]: do we need an account_type enum if it is always just based on validity of openstax_uid?

File: app/models/openstax/accounts/account.rb [hunk 3]
```diff
@@ -46,12 +49,36 @@ def has_authenticated?
       !access_token.nil?
     end
 
+    def valid_openstax_uid?
+      !openstax_uid.nil? && openstax_uid > 0
```
Comment (jpslav) [inline app/models/openstax/accounts/account.rb]: `>=`?

File: db/migrate/7_add_account_type_to_accounts.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class AddAccountTypeToAccounts < ActiveRecord::Migration
+  def change
+    add_column :openstax_accounts_accounts, :account_type, :integer
+
+    OpenStax::Accounts::Account.find_each do |account|
+      account.set_account_type
```
Comment (jpslav) [inline db/migrate/7_add_account_type_to_accounts.rb]: this is going to happen anyway, right?

File: app/models/openstax/accounts/account.rb (hunk 3 above)
Comment (Dantemss) [inline app/models/openstax/accounts/account.rb]: I don't think an ID of 0 would ever happen naturally?

File: app/models/openstax/accounts/account.rb (hunk 2 above)
Comment (Dantemss) [inline app/models/openstax/accounts/account.rb]: Hopefully something that doesn't exist.

File: app/models/openstax/accounts/account.rb (hunk 1 above)
Comment (Dantemss) [inline app/models/openstax/accounts/account.rb]: Could remove account_type and just rely on the openstax_uid being nil or less than 1, but nathan didn't like that.

File: db/migrate/7_add_account_type_to_accounts.rb (hunk 4 above)
Comment (Dantemss) [inline db/migrate/7_add_account_type_to_accounts.rb]: Wouldn't happen automatically for existing accounts until they were updated somewhere else.

**Pull request merged**